Dockerfile

```diff
@@ -1,74 +1,103 @@
+#   2. If buildGradle build argument is set to false, then consumes built Projector assembly from the host.
+#       2.1 Otherwise starts Gradle build of Projector Server and Projector Client.
+#   3. Copies static files to the Projector assembly (entrypoint, launcher, configuration).
+FROM registry.access.redhat.com/ubi8-minimal:8.3-298 as projectorAssembly
+ENV PROJECTOR_DIR /projector
+ENV JAVA_HOME /usr/lib/jvm/java-11
+ARG downloadUrl
+ARG buildGradle
+ADD jetbrains-editor-images/static $PROJECTOR_DIR
+ADD projector-client $PROJECTOR_DIR/projector-client
+ADD projector-server $PROJECTOR_DIR/projector-server
+WORKDIR /downloads
+RUN microdnf install -y wget findutils tar gzip unzip java-11-openjdk-devel \
+    && wget -q $downloadUrl -O - | tar -xz \
+    && find . -maxdepth 1 -type d -name * -exec mv {} $PROJECTOR_DIR/ide \;
+WORKDIR $PROJECTOR_DIR/projector-server
+RUN if [ "$buildGradle" = "true" ]; then ./gradlew clean; else echo "Skipping gradle build"; fi \
+    && if [ "$buildGradle" = "true" ]; then ./gradlew --console=plain :projector-server:distZip; else echo "Skipping gradle build"; fi
+WORKDIR $PROJECTOR_DIR
+RUN cp $PROJECTOR_DIR/projector-server/projector-server/build/distributions/projector-server-1.0-SNAPSHOT.zip $PROJECTOR_DIR \
+    && unzip $PROJECTOR_DIR/projector-server-1.0-SNAPSHOT.zip \
+    && rm $PROJECTOR_DIR/projector-server-1.0-SNAPSHOT.zip \
+    && rm -rf $PROJECTOR_DIR/projector-client \
+    && rm -rf $PROJECTOR_DIR/projector-server \
+    && mv projector-server-1.0-SNAPSHOT $PROJECTOR_DIR/ide/projector-server \
+    && mv $PROJECTOR_DIR/ide-projector-launcher.sh $PROJECTOR_DIR/ide/bin \
+    && chmod 644 $PROJECTOR_DIR/ide/projector-server/lib/*
 
-# Copy sh scripts
-COPY --chown=0:0 scripts/*.sh /opt/
+# Stage 2. Build the main image with necessary environment for running Projector
+#   Doesn't require to be a desktop environment. Projector runs in headless mode.
+FROM registry.access.redhat.com/ubi8-minimal:8.3-298
+ARG downloadUrl
+ENV PROJECTOR_USER_NAME projector-user
+ENV PROJECTOR_DIR /projector
+ENV HOME /home/$PROJECTOR_USER_NAME
+ENV PROJECTOR_CONFIG_DIR $HOME/.config
+RUN set -ex \
+    && microdnf install -y yum \
+    && yum install -y --setopt=tsflags=nodocs \
+    shadow-utils wget git nss procps findutils \
+    # Packages required by JetBrains products.
+    libsecret jq \
+    # Packages needed for AWT.
+    libXext libXrender libXtst libXi freetype \
+    # Check whether download Url represents IntelliJ IDEA to install required dependencies.
+    && if [ "${downloadUrl#*idea}" != "$downloadUrl" ]; then yum install -y java-11-openjdk-devel; else echo "Not IntelliJ Idea"; fi \
+    # Check whether download Url represents PyCharm to install required dependencies.
+    && if [ "${downloadUrl#*pycharm}" != "$downloadUrl" ]; then yum install -y python2 python3 python3-pip python3-setuptools; else echo "Not Pycharm"; fi \
```

Comment:
if already installing `yum` (which depends on & installs python), you might just be able to have one image which contains support for BOTH java11 and python2/3... so that we don't need this if-else block. 

Then you have a simpler Dockerfile that always produces an image that can be used for both pycharm and idea.

Reply:
It is a tricky moment. Because with this Dockerfile it is possible to build for example a WebStorm, which doesn't require neither openjdk nor python. The same for PyCharm, openjdk for it will be redundant. I'm not sure, that packing all the staff into one image, not dividing by IDE type during the build time it is a good idea.